### PR TITLE
[Clang] Fix P2564 handling of variable initializers

### PR DESCRIPTION
**Pull request by katzdm**
The following program produces a diagnostic in Clang and EDG, but compiles correctly in GCC and MSVC:
```cpp
#include <vector>

consteval std::vector<int> fn() { return {1,2,3}; }
constexpr int a = fn()[1];
```

Clang's diagnostic is as follows:
```cpp
<source>:6:19: error: call to consteval function 'fn' is not a constant expression
    6 | constexpr int a = fn()[1];
      |                   ^
<source>:6:19: note: pointer to subobject of heap-allocated object is not a constant expression
/opt/compiler-explorer/gcc-snapshot/lib/gcc/x86_64-linux-gnu/14.0.1/../../../../include/c++/14.0.1/bits/allocator.h:193:31: note: heap allocation performed here
  193 |             return static_cast<_Tp*>(::operator new(__n));
      |                                      ^
1 error generated.
Compiler returned: 1
```

Based on my understanding of [`[dcl.constexpr]/6`](https://eel.is/c++draft/dcl.constexpr#6):
> In any constexpr variable declaration, the full-expression of the initialization shall be a constant expression

It seems to me that GCC and MSVC are correct: the initializer `fn()[1]` does not evaluate to an lvalue referencing a heap-allocated value within the `vector` returned by `fn()`; it evaluates to an lvalue-to-rvalue conversion _from_ that heap-allocated value.

This PR turns out to be a bug fix on the implementation of [P2564R3](https://wg21.link/p2564r3); as such, it only applies to C++23 and later. The core problem is that the definition of a constant-initialized variable ([`[expr.const/2]`](https://eel.is/c++draft/expr.const#2)) is contingent on whether the initializer can be evaluated as a constant expression:

> A variable or temporary object o is _constant-initialized_ if [...] the full-expression of its initialization is a constant expression when interpreted as a _constant-expression_, [...]

That can't be known until we've finished parsing the initializer, by which time we've already added immediate invocations and consteval references to the current expression evaluation context. This will have the effect of evaluating said invocations as full expressions when the context is popped, even if they're subexpressions of a larger constant expression initializer. If, however, the variable _is_ constant-initialized, then its initializer is [manifestly constant-evaluated](https://eel.is/c++draft/expr.const#20):

> An expression or conversion is _manifestly constant-evaluated_ if it is [...] **the initializer of a variable that is usable in constant expressions or has constant initialization** [...]

which in turn means that any subexpressions naming an immediate function are in an [immediate function context](https://eel.is/c++draft/expr.const#16):

> An expression or conversion is in an immediate function context if it is potentially evaluated and either [...] it is a **subexpression of a manifestly constant-evaluated expression** or conversion

and therefore _are not to be considered [immediate invocations](https://eel.is/c++draft/expr.const#16) or [immediate-escalating expressions](https://eel.is/c++draft/expr.const#17) in the first place_:

> An invocation is an _immediate invocation_ if it is a potentially-evaluated explicit or implicit invocation of an immediate function and **is not in an immediate function context**.

> An expression or conversion is _immediate-escalating_ if **it is not initially in an immediate function context** and [...]

The approach that I'm therefore proposing is:
1. Create a new expression evaluation context for _every_ variable initializer (rather than only nonlocal ones).
2. Attach initializers to `VarDecl`s _prior_ to popping the expression evaluation context / scope / etc. This sequences the determination of whether the initializer is in an immediate function context _before_ any contained immediate invocations are evaluated.
3. When popping an expression evaluation context, elide all evaluations of constant invocations, and all checks for consteval references, if the context is an immediate function context. Note that if it could be ascertained that this was an immediate function context at parse-time, we [would never have registered](https://github.com/llvm/llvm-project/blob/760910ddb918d77e7632be1678f69909384d69ae/clang/lib/Sema/SemaExpr.cpp#L17799) these immediate invocations or consteval references in the first place.

Most of the test changes previously made for this PR are now reverted and passing as-is. The only test updates needed are now as follows:
- A few diagnostics in `consteval-cxx2a.cpp` are updated to reflect that it is the `consteval tester::tester` constructor, not the more narrow `make_name` function call, which fails to be evaluated as a constant expression.
- The reclassification of `warn_impcast_integer_precision_constant` as a compile-time diagnostic adds a (somewhat duplicative) warning when attempting to define an enum constant using a narrowing conversion. It also, however, retains the existing diagnostics which @erichkeane (rightly) objected to being lost from an earlier revision of this PR.

### CODE REVIEW

**Comment by cor3ntin** [comment]
I think the change makes sense.
Can you add a release note?

**Comment by katzdm** [comment]
> I think the change makes sense. Can you add a release note?

Appreciate the reminder!

File: clang/test/SemaCXX/cxx2a-consteval.cpp [hunk 1]
```diff
@@ -1032,17 +1043,15 @@ int f() {
...
   };
   constexpr int (*f1)(void) = l1; // expected-error {{constexpr variable 'f1' must be initialized by a constant expression}} \
                                   // expected-note  {{pointer to a consteval declaration is not a constant expression}}
 
 
-  constexpr auto lstatic = []() static consteval { // expected-error {{cannot take address of consteval call operator of '(lambda at}} \
-                                       // expected-note  2{{declared here}} \
-                                       // expected-warning {{extension}}
+  constexpr auto lstatic = []() static consteval { // expected-note  {{declared here}} \
+                                                   // expected-warning {{extension}}
```
[the remark this answers is not in the document]
Comment (katzdm) [reply clang/test/SemaCXX/cxx2a-consteval.cpp]: Not that I can tell; I've added one above which assigns the address of the lambda to a non-constexpr variable. The resulting diagnostic points to the `l1` declaration - seems like ideally it would also diagnose at the site of the initializer, but that appears unrelated to this change.

**Review by katzdm — COMMENTED** [review]
@cor3ntin Thanks for your review; this is ready for another look.

Note that the use of a full `ExpressionEvaluationContext` caused a handful of other tests to start failing - in particular, ones that previously expected warnings related to changes of value from implicit casts. I verified that these warnings are all emitted through `Sema::DiagRuntimeBehavior`, which no-ops when in a `ConstantEvaluated` context - so the lack of warning seems expected since the initializer is now treated as a constant expression.

File: clang/test/AST/Interp/intap.cpp [hunk 2]
```diff
@@ -9,8 +9,7 @@ using MaxBitInt = _BitInt(128);
 
 constexpr _BitInt(2) A = 0;
 constexpr _BitInt(2) B = A + 1;
-constexpr _BitInt(2) C = B + 1; // expected-warning {{from 2 to -2}} \
-                                // ref-warning {{from 2 to -2}}
+constexpr _BitInt(2) C = B + 1;
```
Comment (erichkeane) [inline clang/test/AST/Interp/intap.cpp:12]: Losing this diagnostic is definitely incorrect, this is one we should have kept.

File: clang/test/CXX/expr/expr.const/p2-0x.cpp [hunk 3]
```diff
@@ -244,8 +244,8 @@ namespace UndefinedBehavior {
     constexpr int n13 = n5 + n5; // expected-error {{constant expression}} expected-note {{value -4294967296 is outside the range of }}
     constexpr int n14 = n3 - n5; // expected-error {{constant expression}} expected-note {{value 4294967295 is outside the range of }}
     constexpr int n15 = n5 * n5; // expected-error {{constant expression}} expected-note {{value 4611686018427387904 is outside the range of }}
-    constexpr signed char c1 = 100 * 2; // ok expected-warning{{changes value}}
-    constexpr signed char c2 = '\x64' * '\2'; // also ok  expected-warning{{changes value}}
+    constexpr signed char c1 = 100 * 2; // ok - no error from changing value because initializer is constexpr.
```
Comment (erichkeane) [inline clang/test/CXX/expr/expr.const/p2-0x.cpp:247]: Shouldn't this actually be an error?  Signed overflow during constexpr?

**Review by erichkeane — COMMENTED** [review]
I don't have a great idea on this one, but losing the diagnostics seems incorrect.  I think the whole initializer should probably be treated in 1 shot as a constant expr (so move the ConstantExpr to the top (above the `ImplicitCastExpr` in your example), but we DEFINITELY should be getting some of those diagnostics.

File: clang/test/CXX/expr/expr.const/p2-0x.cpp (hunk 3 above)
Comment (katzdm) [reply clang/test/CXX/expr/expr.const/p2-0x.cpp:247]: @erichkeane I think that it isn't, because `100`, `2`, and `100 * 2` are all expressions of type `int`. The only sideways thing that happens is the implicit conversion from `int` to `signed char` that reinterprets the value as `56`.

Both this diagnostic and the other that you commented on can be recovered by changing [this](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaChecking.cpp#L16543-L16547) `Sema::DiagRuntimeBehavior` to `Sema::Diag` instead.

Running this change with `check-clang`, we recover the following diagnostics that are currently removed by this PR:
- clang/test/AST/Interp/intap.cpp
- clang/test/CXX/expr/expr.const/p2-0x.cpp
- clang/test/SemaCXX/enum-scoped.cpp
- clang/test/SemaCXX/ext-int.cpp

with no other tests newly failing.

Let me know if you think this change makes sense; I'll push it to the PR and revert the changes to the tests mentioned above. Also let me know if it's desirable for any of the other diagnostics removed by this PR to be kept; happy to dig.

Comment (katzdm) [reply clang/test/CXX/expr/expr.const/p2-0x.cpp:247]: Ah, just noticed - `enum-scoped.cpp` isn't a test that's affected by this change as-is. The suggested change adds an additional warning to where there's already an error; the declaration
```cpp
enum : signed char { A = 128 };
```

is diagnosed with:
```cpp
test2.cpp:1:26: warning: implicit conversion from 'int' to 'signed char' changes value from 128 to -128 [-Wconstant-conversion]
    1 | enum : signed char { A = 128 };
      |                          ^~~
test2.cpp:1:26: error: enumerator value evaluates to 128, which cannot be narrowed to type 'signed char' [-Wc++11-narrowing]
```

Not sure how helpful that additional diagnostic is.

File: clang/lib/Parse/ParseDecl.cpp [hunk 4]
```diff
@@ -2554,16 +2554,27 @@ Decl *Parser::ParseDeclarationAfterDeclarator(
...
     Declarator &D;
     Decl *ThisDecl;
+    EnterExpressionEvaluationContext EC;
 
     InitializerScopeRAII(Parser &P, Declarator &D, Decl *ThisDecl)
-        : P(P), D(D), ThisDecl(ThisDecl) {
+        : P(P), D(D), ThisDecl(ThisDecl),
+          EC(P.Actions, Sema::ExpressionEvaluationContext::ConstantEvaluated,
+             ThisDecl, Sema::ExpressionEvaluationContextRecord::EK_Other,
+             isConstexprVariable(ThisDecl)) {
```
Comment (efriedma-quic) [inline clang/lib/Parse/ParseDecl.cpp:2577]: This doesn't look right.  The relevant standard text says:

- "An expression or conversion is in an immediate function context if it is potentially evaluated and [...] it is a subexpression of a manifestly constant-evaluated expression or conversion [...]"
- "An expression or conversion is manifestly constant-evaluated if it is [...] the initializer of a variable that is usable in constant expressions or has constant initialization"

So we need to handle:
- constexpr variables
- variables of type `const int`
- constant-initialized variables (global/static)

Comment (efriedma-quic) [reply clang/lib/Parse/ParseDecl.cpp:2577]: On a side-note, gcc currently accepts the following; if I'm understanding the rule, it's not supposed to be valid.

```
struct S {
  int *value;
  constexpr S(int v) : value(new int {v}) {}
  constexpr ~S() { delete value; }
};
consteval S fn() { return S(5); }
int g();
void f() {
  int a = *fn().value;
}
```

Comment (katzdm) [reply clang/lib/Parse/ParseDecl.cpp:2577]: > if I'm understanding the rule, it's not supposed to be valid.

Agreed; MSVC appears to get this right:
```
<source>(8): error C7595: 'fn': call to immediate function is not a constant expression
```

Since `int a` is not usable in a constant expression, its initializer isn't manifestly constant-evaluated. Thus only the call to `fn()` is manifestly constant-evaluated, so the access to the heap-allocated memory is ill-formed.

...as far as I can tell 😅

Comment (katzdm) [reply clang/lib/Parse/ParseDecl.cpp:2577]: @efriedma-quic I see your point - we need to handle a broader class of variables in this case than just `constexpr`. Let me see what I can do, thanks for the feedback.

**Comment by katzdm** [comment]
This ended up being a lot more subtle than I expected. My latest push proposes a different solution; @cor3ntin and @efriedma-quic -  please let me know your thoughts.

This PR turns out to be a bug fix on the implementation of [P2564R3](https://wg21.link/p2564r3); as such, it only applies to C++23 and later. The core problem is that the definition of a constant-initialized variable ([`[expr.const/2]`](https://eel.is/c++draft/expr.const#2)) is contingent on whether the initializer can be evaluated as a constant expression:

> A variable or temporary object o is _constant-initialized_ if [...] the full-expression of its initialization is a constant expression when interpreted as a _constant-expression_, [...]

That can't be known until we've finished parsing the initializer, by which time we've already added immediate invocations and consteval references to the current expression evaluation context. This will have the effect of evaluating said invocations as full expressions when the context is popped, even if they're subexpressions of a larger constant expression initializer. If, however, the variable _is_ constant-initialized, then its initializer is [manifestly constant-evaluated](https://eel.is/c++draft/expr.const#20):

> An expression or conversion is _manifestly constant-evaluated_ if it is [...] **the initializer of a variable that is usable in constant expressions or has constant initialization** [...]

which in turn means that any subexpressions naming an immediate function are in an [immediate function context](https://eel.is/c++draft/expr.const#16):

> An expression or conversion is in an immediate function context if it is potentially evaluated and either [...] it is a **subexpression of a manifestly constant-evaluated expression** or conversion

and therefore _are not to be considered [immediate invocations](https://eel.is/c++draft/expr.const#16) or [immediate-escalating expressions](https://eel.is/c++draft/expr.const#17) in the first place_:

> An invocation is an _immediate invocation_ if it is a potentially-evaluated explicit or implicit invocation of an immediate function and **is not in an immediate function context**.

> An expression or conversion is _immediate-escalating_ if **it is not initially in an immediate function context** and [...]


The approach that I'm therefore proposing is:
1. Create a new expression evaluation context for _every_ variable initializer (rather than only nonlocal ones).
2. Attach initializers to `VarDecl`s _prior_ to popping the expression evaluation context / scope / etc. This sequences the determination of whether the initializer is in an immediate function context _before_ any contained immediate invocations are evaluated.
3. When popping an expression evaluation context, elide all evaluations of constant invocations, and all checks for consteval references, if the context is an immediate function context. Note that if it could be ascertained that this was an immediate function context at parse-time, we [would never have registered](https://github.com/llvm/llvm-project/blob/760910ddb918d77e7632be1678f69909384d69ae/clang/lib/Sema/SemaExpr.cpp#L17799) these immediate invocations or consteval references in the first place.

Most of the test changes previously made for this PR are now reverted and passing as-is. The only test updates needed are now as follows:
- A few diagnostics in `consteval-cxx2a.cpp` are updated to reflect that it is the `consteval tester::tester` constructor, not the more narrow `make_name` function call, which fails to be evaluated as a constant expression.
- The reclassification of `warn_impcast_integer_precision_constant` as a compile-time diagnostic adds a (somewhat duplicative) warning when attempting to define an enum constant using a narrowing conversion. It also, however, retains the existing diagnostics which @erichkeane (rightly) objected to being lost from an earlier revision of this PR.

Sorry for the long-winded explanation of a relatively small change - the language in `[expr.const]` is (at least I find) quite involved, so I figured more detail couldn't hurt.

File: clang/lib/Parse/ParseDecl.cpp (hunk 4 above)
Comment (katzdm) [reply clang/lib/Parse/ParseDecl.cpp]: @efriedma-quic Your feedback led to my taking a different approach for this PR; please see [this comment](https://github.com/llvm/llvm-project/pull/89565#issuecomment-2079917447).

File: clang/test/CXX/expr/expr.const/p2-0x.cpp (hunk 3 above)
Comment (katzdm) [reply clang/test/CXX/expr/expr.const/p2-0x.cpp]: I think the additional diagnostic is relatively harmless; I've gone ahead with the change I suggested above.

File: clang/test/SemaCXX/cxx2a-consteval.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (katzdm) [reply clang/test/SemaCXX/cxx2a-consteval.cpp]: I've added some test cases for assigning to a non constexpr variable, both within and outside of an immediate function context.

**Comment by efriedma-quic** [comment]
The approach makes sense.  I'll defer to someone who's more familiar with this code for the full review.

File: clang/lib/Parse/ParseDecl.cpp [hunk 5]
```diff
@@ -2559,25 +2559,30 @@ Decl *Parser::ParseDeclarationAfterDeclaratorAndAttributes(
...
-        P.Actions.ActOnCXXEnterDeclInitializer(S, ThisDecl);
+        if (ThisDecl && !ThisDecl->isInvalidDecl()) {
+          P.Actions.ActOnCXXEnterDeclInitializer(S, ThisDecl);
+          Entered = true;
+        }
       }
     }
-    ~InitializerScopeRAII() { pop(); }
-    void pop() {
+    ~InitializerScopeRAII() {
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseDecl.cpp:2578]: This is a nice improvement

File: clang/lib/Sema/SemaChecking.cpp [hunk 6]
```diff
@@ -16548,11 +16548,10 @@ static void CheckImplicitConversion(Sema &S, Expr *E, QualType T,
         std::string PrettySourceValue = toString(Value, 10);
         std::string PrettyTargetValue = PrettyPrintInRange(Value, TargetRange);
 
-        S.DiagRuntimeBehavior(
-            E->getExprLoc(), E,
-            S.PDiag(diag::warn_impcast_integer_precision_constant)
-                << PrettySourceValue << PrettyTargetValue << E->getType() << T
-                << E->getSourceRange() << SourceRange(CC));
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaChecking.cpp:16555]: This seems unrelated - We do not want to emit this diag during constant evaluation, because we know the source value and that it does not lose precision.

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 7]
```diff
@@ -18562,9 +18553,7 @@ static bool isNonlocalVariable(const Decl *D) {
 /// class X. If the declaration had a scope specifier, a scope will have
 /// been created and passed in for this purpose. Otherwise, S will be null.
 void Sema::ActOnCXXEnterDeclInitializer(Scope *S, Decl *D) {
-  // If there is no declaration, there was an error parsing it.
-  if (!D || D->isInvalidDecl())
-    return;
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaDeclCXX.cpp:18567]: This is a bit unrelated. Let's be conservative here

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 8]
```diff
@@ -18573,25 +18562,35 @@ void Sema::ActOnCXXEnterDeclInitializer(Scope *S, Decl *D) {
...
-        ExpressionEvaluationContext::PotentiallyEvaluated, D);
+  PushExpressionEvaluationContext(
+      ExpressionEvaluationContext::PotentiallyEvaluated, D);
 }
 
 /// Invoked after we are finished parsing an initializer for the declaration D.
 void Sema::ActOnCXXExitDeclInitializer(Scope *S, Decl *D) {
-  // If there is no declaration, there was an error parsing it.
-  if (!D || D->isInvalidDecl())
-    return;
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaDeclCXX.cpp:18588]: Same here

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 9]
```diff
@@ -18573,25 +18562,35 @@ void Sema::ActOnCXXEnterDeclInitializer(Scope *S, Decl *D) {
...
+      // An expression or conversion is in an 'immediate function context' if it
+      // is potentially evaluated and either:
+      // [...]
+      // - it is a subexpression of a manifestly constant-evaluated expression
+      //   or conversion.
+      ExprEvalContexts.back().InImmediateFunctionContext = true;
+    }
+  }
+
+  PopExpressionEvaluationContext();
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaDeclCXX.cpp:18593]: I would like to see a comment here because it is very subtle.
We parse the whole thing as a not constant expression, so immediate invocation gets delayed,
and we want to evaluate them when the variable expression context is popped, unless the variable is consteval (which imply we have always 2 expression contexts)

File: clang/lib/Sema/SemaDeclCXX.cpp (hunk 7 above)
Comment (katzdm) [reply clang/lib/Sema/SemaDeclCXX.cpp:18567]: Ah, I see how it could look that way; it's not in fact unrelated.

Prior this PR, `InitializerScopeRAII::pop()` was called _before_ attaching the initializer to the `VarDecl`. This gave an invariant that if the `VarDecl` was valid when the scope was entered, it would continue to be valid when it was exited.

With this PR, the `VarDecl` can invalidated when the initializer is attached; `D->isInvalidDecl()` might be false when calling `ActOnCXXEnterDeclInitializer`, but might be true when calling `ActOnCXXExitDeclInitializer`. We therefore can't just no-op in `ActOnCXXExitDeclInitializer` when the variable is invalid, or else we'll never pop the expression evaluation context (leading to assert-failures from scope entity mismatch, etc).

This is what motivates adding an `Entered` field to `InitializerScopeRAII` (`false` if `!D` or `D->isInvalidDecl` at scope-entry time) to remember whether `ActOnCXXExitDeclInitializer` needs to be called at scope-exit time. So this check is still present, but it's been lifted from this callsite to `InitializerScopeRAII`; I figured it couldn't hurt to leave some assertions to ensure the expected preconditions continue to be met going forward.

Let me know if that seems reasonable, or if there's another way you'd rather see this done.

File: clang/lib/Sema/SemaChecking.cpp (hunk 6 above)
Comment (katzdm) [reply clang/lib/Sema/SemaChecking.cpp:16555]: I'm fine either way, but the following three warning diagnostics are lost without this change:
- [CXX/expr/expr.const/p2-0x.cpp#L247](https://github.com/llvm/llvm-project/blob/main/clang/test/CXX/expr/expr.const/p2-0x.cpp#L247)
- [CXX/expr/expr.const/p2-0x.cpp$248](https://github.com/llvm/llvm-project/blob/main/clang/test/CXX/expr/expr.const/p2-0x.cpp#L248)
- [AST/Interp/intap.cpp#L12](https://github.com/llvm/llvm-project/blob/main/clang/test/AST/Interp/intap.cpp#L12)

I think @erichkeane previously objected to losing these diagnostics.

File: clang/lib/Sema/SemaDeclCXX.cpp (hunk 9 above)
Comment (katzdm) [reply clang/lib/Sema/SemaDeclCXX.cpp]: I've added a clarifying comment - let me know if you think it's sufficient.

File: clang/lib/Sema/SemaChecking.cpp (hunk 6 above)
Comment (cor3ntin) [reply clang/lib/Sema/SemaChecking.cpp:16555]: So, this is interesting.
The standard does not require the conversion to be ill-formed, so your change is actually correct.

File: clang/lib/Sema/SemaDeclCXX.cpp (hunk 7 above)
Comment (cor3ntin) [reply clang/lib/Sema/SemaDeclCXX.cpp:18567]: I see, make sense. Thanks

File: clang/lib/Sema/SemaChecking.cpp (hunk 6 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaChecking.cpp:16555]: Its a warning, of course the standard doesn't require it.  It seems to me that it is a USEFUL diagnostic though.

Comment (katzdm) [reply clang/lib/Sema/SemaChecking.cpp:16555]: @cor3ntin Do you think any changes to diagnostics need to happen in this PR, or are you okay with the warning applying during constant evaluation?

Comment (cor3ntin) [reply clang/lib/Sema/SemaChecking.cpp:16555]: I'm fine with things as-is.
As a follow up we could: 
 - Make it a warning that default to an error
 - Try to remove the duplication in C and in the `constexpr char c = {1000}` case

**Comment by katzdm** [comment]
👋 Friendly ping - No worries if more time is needed to consider the impact of this one, but wanted to check if any more changes are needed at this time.

**Review by erichkeane — APPROVED** [review]
This seems reasonable to me, I think @cor3ntin should also approve before committing though.

**Review by cor3ntin — APPROVED** [review]
LGTM too

**Review by Endilll — COMMENTED** [review]
`Sema.h` changes look good.

**Comment by cor3ntin** [comment]
CI failures look unrelated

**Comment by Fznamznon** [comment]
@katzdm 
It seems after this a `warn_impcast_integer_precision_constant` warning went missing:

https://godbolt.org/z/ndsPb44b4

Is that expected?

**Comment by katzdm** [comment]
@Fznamznon  Hmm...my suggested change does seem to break some tests :( But to circle back to your original question, I think it's not totally unexpected - a broader class of variable initializers are now (correctly afaict) evaluated as constant expressions, which could make some warnings marked as "runtime behavior" no longer apply. I agree though that the result is unfortunate - @erichkeane had pushed back on some other warnings similarly getting swallowed during review.

I wonder if we're taking the wrong approach here - we could also leave both of these warnings as "runtime only", and separately try to diagnose during constant evaluation. If that could work, it might make it easier to e.g., only diagnose overflow in a branch of a compile-time conditional if it's actually evaluated (which GCC seems to manage to do).

**Comment by efriedma-quic** [comment]
clang already generates certain diagnostics from ExprConstant; expanding the set of diagnostics could be reasonable.

Alternatively, we could try to add some sort of cooperation between DiagRuntimeBehavior and constant evaluation, to try to avoid having to diagnose everything twice: basically, Sema makes a list of constant-evaluated expressions it wants to warn about, then if ExprConstant prints a warning if it sees one of those expressions.

**Comment by Fznamznon** [comment]
@katzdm 
Does it make sense to file an issue so we don't forget to bring the warning back?

**Comment by katzdm** [comment]
> @katzdm Does it make sense to file an issue so we don't forget to bring the warning back?

Yep, for sure - I've opened #92656.
